src/tpm2.rs

```diff
@@ -0,0 +1,957 @@
+                self.get_tpmdata("aik_pw"),
+                self.get_tpmdata("owner_pw"),
+            ),
+        };
+
+        self.run(cmd, Some(vec![tf_sec_path]))
+            .and_then(|(_, f_out)| match f_out.get(tf_sec_path) {
+                None => Err(KeylimeTpmError::new_tpm_rust_error(
+                    "Error: invalid file output.",
+                )),
+                Some(content) => Ok(base64::encode(content)),
+            })
+    }
+
+    fn run(
+        &self,
+        command: String,
+        output_path: Option<Vec<&str>>,
+    ) -> Result<(String, HashMap<String, String>), KeylimeTpmError> {
+        let words: Vec<&str> = command.split(" ").collect();
+        let mut number_tries = 0;
+        let args = &words[1..words.len()];
+        let cmd = &words[0];
+        let mut env_vars: HashMap<String, String> = HashMap::new();
+        for (key, value) in env::vars() {
+            env_vars.insert(key.to_string(), value.to_string());
+        }
+        let lib_path = env_vars
+            .get("LD_LIBRARY_PATH")
+            .map_or_else(|| String::new(), |v| v.clone());
+        env_vars.insert(
+            "LD_LIBRARY_PATH".to_string(),
+            format!("{}:{}", lib_path, common::TPM_LIBS_PATH),
+        );
+        env_vars.insert(
+            "TPM2TOOLS_TCTI".to_string(),
+            "tabrmd:bus_name=com.intel.tss2.Tabrmd".to_string(),
+        );
+        match env_vars.get_mut("PATH") {
+            Some(v) => v.push_str(common::TPM_TOOLS_PATH),
+            None => {
+                return Err(KeylimeTpmError::new_tpm_rust_error(
+                    "PATH envrionment variable dosen't exist.",
+                ));
+            }
+        }
+        let mut output: Output;
+        'exec: loop {
+            let t0 = SystemTime::now();
+
+            output =
+                Command::new(&cmd).args(args).envs(&env_vars).output()?;
+
+            let t_diff = t0.duration_since(t0)?;
+            info!("Time cost: {}", t_diff.as_secs());
+            match output.status.code() {
+                Some(TPM_IO_ERROR) => {
+                    number_tries += 1;
+                    if number_tries >= MAX_TRY {
+                        return Err(KeylimeTpmError::new_tpm_error(
+                            TPM_IO_ERROR,
+                            format!(
+                                "{}{}{}{}",
+                                "TPM appears to be in use by another ",
+                                "application. Keylime is incompatible with ",
+                                "with other TPM TSS application like trousers/",
+                                "tpm-tools. Please uninstall or disable.",
+                            )
+                            .as_str(),
+                        ));
+                    }
+
+                    info!(
+                        "Failed to call TPM {}/{} times, try again in {}s.",
+                        number_tries, MAX_TRY, RETRY,
```

Comment:
I can't work out what `RETRY` is for, its set as a const and mutated anywhere else..it will always be `4` from what I can see?

Reply:
Is this meant to be `retry_interval`?

https://github.com/keylime/keylime/blob/adef1c831f858429560b292f9eebd91c37cc776f/keylime.conf#L101-L103

Reply:
Further reading, I guess this is meant to be `RETRY_SLEEP`

Is it possible we could use the same naming conventions set in the keylime.conf file.

I have a PR ready to implement INI config loading, and so it makes it easier for me to map values.

Reply:
That is a good catch!! Sorry for the confusion.